packages/stack-shared/src/config/parser.ts

```
@@ -0,0 +1,163 @@
+// Check out https://github.com/stack-auth/info/blob/main/eng-handbook/random-thoughts/config-json-format.md for more information on the config format
+
+import * as yup from "yup";
+import { yupNumber, yupObject, yupTuple } from "../schema-fields";
+import { deepPlainClone } from "../utils/objects";
+
+export function mergeConfigs(options: {
+  configs: any[],
+  configSchema: yup.AnySchema,
+}) {
+  let mergedConfig: {} | yup.InferType<typeof options.configSchema> = {};
+
+  for (const config of options.configs) {
+    mergedConfig = mergeConfig({
+      configSchema: options.configSchema,
+      defaultConfig: mergedConfig,
+      overrideConfig: config,
+    });
+  }
+
+  return mergedConfig;
+}
+
+export function mergeConfig(options: {
+  configSchema: yup.AnySchema,
+  defaultConfig: any,
+  overrideConfig: any,
+}): yup.InferType<typeof options.configSchema> {
+  const newConfig = deepPlainClone(options.defaultConfig) as any;
+
+  Object.keys(options.overrideConfig).forEach((key) => {
+    const overrideValue = options.overrideConfig[key];
+    const pathParts = key.split('.');
+    let target = newConfig;
+
+    // Traverse down to the parent of the final key.
+    for (let i = 0; i < pathParts.length - 1; i++) {
+      const part = pathParts[i];
+      // Check if part is a valid array index
+      const arrayIndex = parseInt(part);
+      if (!isNaN(arrayIndex) && Array.isArray(target)) {
+        // Handle array index
+        if (!target[arrayIndex]) {
+          target[arrayIndex] = {};
+        }
+        target = target[arrayIndex];
+      } else if (typeof target[part] !== 'object' || target[part] === null) {
+        target[part] = {};
```

we should throw an error instead